Source/WebCore/automation/AutomationInstrumentation.cpp

```diff
@@ -0,0 +1,111 @@
+* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "config.h"
+#include "AutomationInstrumentation.h"
+
+#if ENABLE(WEBDRIVER_BIDI)
+
+#include <JavaScriptCore/ConsoleMessage.h>
+#include <JavaScriptCore/ConsoleTypes.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/Observer.h>
+#include <wtf/StdLibExtras.h>
+#include <wtf/WeakPtr.h>
+
+namespace WebCore {
+
+using namespace Inspector;
+
+namespace {
+static WeakPtr<AutomationInstrumentationClient>& automationClient()
+{
+    static NeverDestroyed<WeakPtr<AutomationInstrumentationClient>> s_client;
+    return s_client.get();
+}
+}
+
+void AutomationInstrumentation::setClient(const AutomationInstrumentationClient &client)
+{
+    ASSERT(!automationClient());
+    automationClient() = &client;
+}
+
+void AutomationInstrumentation::clearClient()
+{
+    automationClient().clear();
+}
+
+static String logEntryLevelForMessage(const ConsoleMessage& message)
+{
+    if (message.type() == MessageType::Assert || message.level() == MessageLevel::Error)
+        return "error"_s;
+    if (message.type() == MessageType::Trace || message.level() == MessageLevel::Debug)
+        return "debug"_s;
+    if (message.level() == MessageLevel::Warning)
+        return "warn"_s;
+    return "info"_s;
+}
+
+static String logEntryMethodNameForMessage(const ConsoleMessage& message)
+{
+    if (message.type() == MessageType::Assert)
+        return "assert"_s;
+    if (message.type() == MessageType::Trace)
+        return "trace"_s;
+    if (message.level() == MessageLevel::Warning)
+        return "warn"_s;
+    if (message.level() == MessageLevel::Error)
+        return "error"_s;
+    if (message.level() == MessageLevel::Debug)
+        return "debug"_s;
+    return "log"_s;
+}
+
+void AutomationInstrumentation::addMessageToConsole(const std::unique_ptr<ConsoleMessage>& message)
+{
+    if (LIKELY(!automationClient()))
+        return;
+
+    auto level = logEntryLevelForMessage(*message);
+    auto text = message->message();
+    // FIXME Check whether we're getting this as an JS error, so type should be "javascript"
+    auto type = "console"_s;
+    auto timestamp = message->timestamp();
+    auto method = logEntryMethodNameForMessage(*message);
+    WTF::ensureOnMainThread([level, text, type, timestamp, method] {
+        if (LIKELY(automationClient()))
+            automationClient()->addMessageToConsole(level, text, type, timestamp, method);
```

Comment:
i think it actually makes more sense to send the typed values instead of strings since it's possible some other client might want to do things with them in code
```cpp
WTF::ensureOnMainThread([source = message.source(), type = message.type(), level = message.level(), text = message.message(), timestamp = message.timestamp()] {
    if (LIKELY(automationClient()))
        automationClient()->addMessageToConsole(source, type, level, text, timestamp);
});
```
the client should be the one responsible for converting the typed values into strings, and probably only right before it puts them into the JSON since it's much cheaper/easier/etc. to send typed values over IPC than strings